parachain/src/wasm_executor/workspace.rs

```diff
@@ -0,0 +1,523 @@
+			let result_ready_ev = add_event(base_ptr, &mut consumed, mode);
+			let worker_ready_ev = add_event(base_ptr, &mut consumed, mode);
+
+			// The size of AtomicBool is guaranteed to be the same as the bool, however, docs
+			// on the bool primitve doesn't actually state that the in-memory size is equal to 1 byte.
+			//
+			// AtomicBool requires hardware support of 1 byte width of atomic operations though, so
+			// that should be fine.
+			//
+			// We still assert here to be safe than sorry.
+			static_assertions::assert_eq_size!(AtomicBool, u8);
+
+			// SAFETY: `AtomicBool` is represented by an u8 thus will be happy to take any alignment.
+			let attached = base_ptr.add(consumed) as *mut AtomicBool;
+
+			let consumed = align_up_to(consumed, 64);
+
+			Self {
+				shmem,
+				attached,
+				consumed,
+				candidate_ready_ev,
+				result_ready_ev,
+				worker_ready_ev,
+			}
+		}
+	}
+
+	fn as_slice(&self) -> &[u8] {
+		unsafe {
+			let base_ptr = self.shmem.as_ptr().add(self.consumed);
+			let remaining = self.shmem.len() - self.consumed;
+			slice::from_raw_parts(base_ptr, remaining)
+		}
+	}
+
+	fn as_slice_mut(&mut self) -> &mut [u8] {
+		unsafe {
+			let base_ptr = self.shmem.as_ptr().add(self.consumed);
+			let remaining = self.shmem.len() - self.consumed;
+			slice::from_raw_parts_mut(base_ptr, remaining)
+		}
+	}
+
+	/// Mark that this workspace has attached. Returns `true` if this instance was attached the
+	/// first.
+	fn declare_exclusive_attached(&self) -> bool {
+		unsafe {
+			// If this succeeded then the value was `false`, thus, we managed to attach exclusively.
+			(&*self.attached)
+				.compare_exchange_weak(
+					false,
+					true,
+					std::sync::atomic::Ordering::SeqCst,
+					std::sync::atomic::Ordering::SeqCst,
+				)
+				.is_ok()
+		}
+	}
+}
+
+fn align_up_to(v: usize, alignment: usize) -> usize {
+	(v + alignment - 1) & !(alignment - 1)
+}
+
+/// Initializes a new or attaches to an exising event.
+///
+/// # Safety
+///
+/// This function should be called with the combination of `base_ptr` and `consumed` so that `base_ptr + consumed`
+/// points on the memory area that is allocated and accessible.
+///
+/// This function should be called only once for the same combination of the `base_ptr + consumed` and the mode.
+/// Furthermore, this function should be called once for initialization.
+///
+/// Specifically, `consumed` should not be modified by the caller, it should be passed as is to this function.
+unsafe fn add_event(base_ptr: *mut u8, consumed: &mut usize, mode: Mode) -> Box<dyn EventImpl> {
+	// SAFETY: there is no safety proof since the documentation doesn't specify the particular constraints
+	//         besides requiring the pointer to be valid. AFAICT, the pointer is valid.
+	let ptr = base_ptr.add(*consumed);
+	let (ev, used_bytes) = match mode {
+		Mode::Initialize => Event::new(ptr, true).unwrap(),
+		Mode::Attach => Event::from_existing(ptr).unwrap(),
```

Comment:
Remove `unwrap` here in favor of `expect`?

Reply:
I missed it 🤦